include/ghex/unstructured/communication_object_ipr.hpp

```diff
@@ -0,0 +1,561 @@
+                            auto& q = tag_offsets_and_max_tags_per_id[domain_ids[i]];
+                            if (q.empty()) {
+                                q.push(std::make_pair(0, max_tags[i]));
+                            } else {
+                                q.push(std::make_pair(q.back().first + q.back().second + 1, max_tags[i]));
+                            }
+                        }
+                        int tag_offsets[sizeof...(Fields)];
+                        for (std::size_t i = 0; i < sizeof...(Fields); ++i) {
+                            auto& q = tag_offsets_and_max_tags_per_id[domain_ids[i]];
+                            tag_offsets[i] = q.front().first;
+                            q.pop();
+                        }
+
+                        // store arguments and corresponding memory in tuples
+                        using buffer_infos_ptr_t = std::tuple<decltype(buffer_infos)*...>;
+                        using memory_t = std::tuple<buffer_memory<Archs>*...>;
+                        buffer_infos_ptr_t buffer_info_tuple{&buffer_infos...};
+                        memory_t memory_tuple{&(std::get<buffer_memory<Archs>>(m_mem))...}; // pointers to the buffer memory for the corresponding device
+
+                        // loop over memory / buffer_infos and compute required space
+                        int i = 0;
+                        detail::for_each(memory_tuple, buffer_info_tuple, [this, &i, &tag_offsets](auto mem, auto bi) {
+                            using arch_type = typename std::remove_reference_t<decltype(*mem)>::arch_type;
+                            using value_type = typename std::remove_reference_t<decltype(*bi)>::value_type;
+                            auto field_ptr = &(bi->get_field());
+                            const domain_id_type my_dom_id = bi->get_field().domain_id();
+                            allocate<arch_type,value_type>(mem, bi->get_pattern(), field_ptr, my_dom_id, bi->device_id(), tag_offsets[i]);
+                            ++i;
+                        });
+                        handle_type h([this](){ this->wait(); });
+                        post_recvs();
+                        pack();
+                        return h;
+                    }
+
+                    void post_recvs() {
+                        detail::for_each(m_mem, [this](auto& m) {
+                            for (auto& p0 : m.recv_memory) {
+                                for (auto& p1: p0.second) {
+                                    if (p1.second.size > 0u) {
+                                        p1.second.message.resize(p1.second.size);
+                                        m.m_recv_futures.emplace_back(
+                                            typename std::remove_reference_t<decltype(m)>::future_type{
+                                                m_comm.recv(p1.second.message, p1.second.address, p1.second.tag).m_handle
```

Comment:
Here I used the same constructor as in `communication_object_2`, i.e. the one with `handle` as argument, which moves only the handle. Not a big deal, but I think that using directly the future move constructor could also be an option, since we are using the same future specialization in both cases anyway. Probably this could be a bit too strict if one day we should decide (for some reason) not to use the `future<void>` specialization as return type for the send / recv functions.

Reply:
I'm not sure I understand completely. The trade-off between reuse of patterns and code is tricky. I would not spend too much thinking about possible future choices, and maybe put a comment explaining what to do if an error appears about it

Reply:
Ok, I just added a comment in the code